rust-runtime/aws-smithy-client/src/retry/adaptive.rs

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Adaptive Retry Strategy
+
+use crate::retry::{NewRequestPolicy, DEFAULT_MAX_ATTEMPTS};
+use aws_smithy_async::rt::sleep::AsyncSleep;
+use aws_smithy_http::operation::Operation;
+use aws_smithy_http::result::{SdkError, SdkSuccess};
+use aws_smithy_http::retry::ClassifyRetry;
+use aws_smithy_types::retry::{ErrorKind, RetryKind};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
```

Comment:
We use `std::sync::Mutex` as opposed to `tokio::sync::Mutex` is because the latter is tied to a specific async runtime construct and we want to avoid that?

Reply:
No. `tokio`'s Mutex actually works fine on any async runtime. The real issue is that locking it is async and our current retry code is all sync.